### ISSUE TITLE
Make `Eq` constraint on store, props, state optional?  Use stable names for equality performance?

### ISSUE THREAD

**Issue by fisx** [liqula/react-hs#8]
As suggested here: https://bitbucket.org/wuzzeb/react-flux/pull-requests/9/use-eq-instead-of-comparing-generated-js/diff#comment-34035864

I only want to do this if there is a reasonable benchmark that shows that it makes a difference in practice.

**Comment by fisx** [comment]
@tysonzero writes in #49:

> So I want to pass functions into child views, which I was able to do in react.js without any difficulty, but I cannot do this unless the Eq constraint on props is removed.

Just curious: how do you decide whether `shouldComponentUpdate` in js if you do not have equality on the props?

I am very interested in this problem.  I can see two solutions that I suspect you won't like:

1. If you have a fixed, small number of lenses that you want to insert into the type, you can define a DSL, i.e. an ADT with one constructor for each lens, and write an interpreter function that replaces those constructors with the lenses in the place where you want to call them.
1. If different lens values alone do not suggest a re-render, you can define your own `Eq` instance that omits comparing them.  (You may be able to add a field to your datatype that changes when re-rendering is required in a way that can be caught by your `Eq` instance.)

> I'm assuming it is an optimization to try and avoid unnecessary rerendering.

Correct.  there is a default `shouldComponentUpdate` that checks old props against new props.

You are also right to worry about making things worse this way.

> I propose instead using either reallyUnsafePtrEquality# or unsafePerformIO combined with stable names.

Can you get this to work from iniside a custom `Eq` instance for your type?  If not, why?

I'd be interested to see your code, both for the solution of the `Eq` problem (if/when you have picked and implemented one) and for the way you are avoiding `mkStatefulView`.  Is it open source?  If not, can you quote bits from it and add it to the react-hs-examles packages as pseudo-code?

**Comment by tysonzero** [comment]
> Just curious: how do you decide whether shouldComponentUpdate in js if you do not have equality on the props?

I mean that is what the pointer / reference equality stuff is for. I mean by default in react.js that function always returns true, and the docs seem to suggest only really deviating from that if you are having performance issues with a specific component.

Honestly the more I look at the native react.js setup the more convinced I am that deep equality checks are not the way to go. Since even if you inherit from `PureComponent` or whatever in react for optimization it still only does a shallow comparison.

Particularly since as I said the whole "only viewing part of a prop's data" risk of pessimization means it isn't actually a safe optimization.

> I am very interested in this problem. I can see two solutions that I suspect you won't like:

> 1. If you have a fixed, small number of lenses that you want to insert into the type, you can define a DSL, i.e. an ADT with one constructor for each lens, and write an interpreter function that replaces those constructors with the lenses in the place where you want to call them.

An ADT like solution would be incompatible with the whole easy composeability stuff which was pretty much the primary goal of this. With my desired setup I can just add the relevant piece of state and a child view to the relevant state and parent view and pass in the lens, and it all just works.

I even thought of some sort of weird existential type + typeable stuff to try and make that composeable, and no matter what I came up with it still involved a LOT more boilerplate, and was all around a dirty hack.

> 2. If different lens values alone do not suggest a re-render, you can define your own Eq instance that omits comparing them. (You may be able to add a field to your datatype that changes when re-rendering is required in a way that can be caught by your Eq instance.)

If a key of an item changes the lens would change but the value wouldn't, but a rerender would still be a good idea.

> Can you get this to work from iniside a custom Eq instance for your type? If not, why?

I mean that `Eq` instance would be not even remotely law abiding and would be downright dangerous to have lying around. It's only safe to use that `unsafe...` stuff in the very limited internal code that deals with rerendering.

And with all the above points I am begging you to please not make me do that.

Currently I actually have the orphan instance `instance Eq (a -> b) where _ == _ = False` which is evil, but I really wanted to make sure this approach was worth going for.

Its worth noting that I am absolutely loving this approach, it is extremely composeable and deals with pretty much all the things I have complained about to you in the past, such as the form errors example, and having what is essentially true local state.

> I'd be interested to see your code, both for the solution of the Eq problem (if/when you have picked and implemented one) and for the way you are avoiding mkStatefulView. Is it open source? If not, can you quote bits from it and add it to the react-hs-examles packages as pseudo-code?

So the code itself for the `Eq` problem is actually very simple, remove the `Eq` constraint from all the relevant places, then write:

    unsafeEq :: a -> a -> Bool
    unsafeEq = eqStableName `on` unsafePerformIO . makeStableName

And every time you previously called `==` now call `unsafeEq`. I named it `unsafeEq` since it is not safe to use in general code, only for stuff like this where false negatives aren't a big deal.

As for avoiding the `mkStatefulView` stuff I can absolutely upload my minimal example. It's actually a fully setup project where you can run `./build.sh` to make the index file render the whole project, or call `stack repl` to interact with the code. The only quirk is the whole `instance Eq (a -> b)` orphan that makes it currently unviable for production code.

Where would you like me to put it? If we want to make subdirectories in react-hs-examples for different examples (and move the current example into one of those subdirectories). I'd be more than happy to create a subdirectory for this minimal example.

**Comment by tysonzero** [comment]
If you accept the pull request I made that moves the todo example into a subdirectory I will submit another PR with the example I talked about here.

**Comment by tysonzero** [comment]
Another thing is you could also add something like `mkViewWithOpts` that takes in a bunch of options including `shouldComponentUpdate`. But I am absolutely positive that defaulting to not doing any `Eq` comparisons is the way to go, but doing some stable name comparisons is not a bad idea, as that lacks the pessimization risk and usability issues.

Note that react.js does it this way, defaulting to not doing any comparisons and allows you to define you own `shouldComponentUpdate` function for when you need to optimize.

One of my biggest goals for `react-hs` is making sure that it is just as / more usable than react.js. And the two main things right now that I can see deviating from this are requiring `Eq` on props (which react.js does not) and disallowing `IO` in handlers (which react.js allows). At the moment I care more about the former since the redux style approach makes the second not really affect me personally, although I do think it is still worth more discussion.

**Comment by tysonzero** [comment]
If I implemented this would you accept the PR? It's currently kind of a blocker, as I want to use the redux approach for further development but cannot in good conscience put `instance Eq (a -> b)` in production.

**Comment by fisx** [comment]
I would like to have a way to pick trivially between:

1. `shouldComponentUpdate = () => true;`,
2. the default we're currently running, and
3. something like 2, but with stable names.

Note that `mkLifecycleView` has callbacks already, so we could arguably just make everybody use that when they really want `shouldComponentUpdate`.

You have a good case here, and I have talked to other people asking for similar changes, so I would really like to see a PR.  And I'll try very hard to accept it.  (-:

Btw do you have you done any benchmarking related to this?

**Comment by fisx** [comment]
> So the code itself for the `Eq` problem is actually very simple, remove the `Eq` constraint from all the relevant places, then write:
> 
>     unsafeEq :: a -> a -> Bool
>     unsafeEq = eqStableName `on` unsafePerformIO . makeStableName
> 
> And every time you previously called `==` now call `unsafeEq`. I named it `unsafeEq` since it is not safe to use in general code, only for stuff like this where false negatives aren't a big deal.

Yes, I think this is a good solution as long as (a) the implementation of `unsafeEq` explains that it sometimes wrongly returns `False`, and why it's ok to use that as you propose; and (b) we add `shouldComponentUpdate` to `LifecycleViewConfig`.

Ideally, `LifecycleViewConfig` would be somehow accepted by all view types, like this:

```haskell
registerCallbacks :: LifecycleViewConfig -> View _ -> View _
```

then we could finally get rid of the Outdated module (#7).

I think I finally read through all your comments now :-).

**Comment by tysonzero** [comment]
Is it alright if for now I just implement the stablenames or `const True` solution and after that is merged we work on the more configurable stuff? I do agree some sort of `ViewConfig` type that can be attached to a `View` would be nice.

Should we default to stable names or to nothing at all? React.js defaults to nothing at all but it seems pretty safe to default to either (right?). I'd be happy with implementing either, all I really care about at the moment is being able to use the redux approach in prod, and in general about being able to do all the stuff I could do in react.js without some kind of `instance Eq (a -> b)` hack.

**Comment by tysonzero** [comment]
I have not done any benchmarking. I think it is safe to implement `const True` / stablenames though since react.js does the former by default, and the latter takes very few extra cpu cycles so it should also be a safe default (unless I'm missing something). But later once we have `ViewConfig` or similar then we should try out some benchmarks for sure.

**Comment by fisx** [comment]
Ok, if you give us a patch for stable names instead of `Eq`, we will run it against our application.  If we can't find any problems there, we can merge it without the changes to the lifecycle view code.

**Comment by tysonzero** [comment]
Awesome. Just submitted the pull request: #62.

**Comment by fisx** [comment]
Haven't forgotten about you, but this week won't see much input from me here.  I am convinced that there should be a way to accomodate you and get rid of the mandatory `Eq` constraint, but getting this into master will probably have to wait until next week.  Sorry-
